src/check/check_types.zig

```diff
@@ -226,102 +227,92 @@ pub fn checkExpr(self: *Self, expr_idx: CIR.Expr.Idx) std.mem.Allocator.Error!bo
-                    // The unification will propagate the actual return type to the call
-                    //
-                    // TODO: Do we need to insert a CIR placeholder node here as well?
-                    // What happens if later this type variable has a problem, and we
-                    // try to look up it's region in CIR?
-                    const func_content = self.types.mkFuncUnbound(arg_vars, call_var);
-                    const expected_func_var = self.types.freshFromContent(func_content);
-                    _ = self.unify(expected_func_var, func_var);
+                        .fn_unbound => |_| {
+                            if (self.types.needsInstantiation(current_func_var)) {
+                                const instantiated_var = try instantiate.instantiateVar(self.types, current_func_var, self.gpa);
+                                const resolved_inst = self.types.resolveVar(instantiated_var);
+                                std.debug.assert(resolved_inst.desc.content == .structure);
+                                std.debug.assert(resolved_inst.desc.content.structure == .fn_unbound);
+                                const inst_func = resolved_inst.desc.content.structure.fn_unbound;
+                                try self.unifyCallWithFunc(call_var, inst_func, call_args, func_var);
+                                return does_fx;
+                            }
+                        },
+                        else => {
+                            // Non-function structure - fall through
+                        },
+                    },
+                    .alias => |alias| {
+                        // Resolve the alias, then continue on to the appropriate branch.
+                        // (It might be another alias, or we might be done and ready to proceed.)
+                        const backing_var = alias.getBackingVar(current_func_var);
+                        current_func_var = backing_var;
+                        current_content = self.types.resolveVar(backing_var).desc.content;
+                        continue :content_switch current_content;
+                    },
+                    else => {
+                        // Non-structure content - fall through
+                    },
                 }
+
+                // We didn't handle the function call above (either because it wasn't a function
+                // or it didn't need instantiation), so fall back on this logic.
+                const arg_vars: []Var = @constCast(@ptrCast(@alignCast(call_args)));
+
+                // Create an unbound function type with the call result as return type
+                // The unification will propagate the actual return type to the call
+                //
+                // TODO: Do we need to insert a CIR placeholder node here as well?
+                // What happens if later this type variable has a problem, and we
+                // try to look up its region in CIR?
```

Comment:
I ran into this too. My solution was to insert placeholder CIR nodes, passing in the region. See [here](https://github.com/roc-lang/roc/pull/7953/files#diff-cb0b4cf3919b58bef567455177f190aafd5b9a4065f99ba3e6914bc47ac62afdR415) and [here](https://github.com/roc-lang/roc/pull/7953/files#diff-9ce821ce803141131d6c5fc376157b33681bec02882a64e0dfd7b921f0c26f81R2743)